keras_nlp/models/gpt2/gpt2_causal_lm.py

```diff
@@ -248,30 +252,76 @@ def call_with_cache(self, token_ids, padding_mask, cache, cache_index):
         )
         return x, cache
 
-    def build_empty_cache(self, batch_size, max_length):
+    def build_cache(self, prompt):
         """Build an empty cache for use with `call_with_cache()`."""
+        batch_size, max_length = tf.shape(prompt)[0], tf.shape(prompt)[1]
         num_layers = self.backbone.num_layers
         num_heads = self.backbone.num_heads
         head_dim = self.backbone.hidden_dim // self.backbone.num_heads
         shape = [batch_size, num_layers, 2, max_length, num_heads, head_dim]
-        return tf.zeros(shape)
+        cache = tf.zeros(shape)
+        # Seed the cache.
+        _, cache = self.call_with_cache(prompt, cache, 0)
+        return cache
 
-    def _get_token_probability(
+    def compile(
         self,
-        prompt,
-        mask,
-        cache=None,
-        cache_index=None,
+        *args,
+        run_eagerly=False,
+        jit_compile=True,
+        sampler="top_k",
+        **kwargs,
     ):
-        batch_size = tf.shape(prompt)[0]
-        prompt = tf.slice(prompt, [0, cache_index], [batch_size, 1])
-        return self.call_with_cache(prompt, mask, cache, cache_index)
+        xla_compatible = is_xla_compatible(self)
+        super().compile(
+            *args,
+            run_eagerly=run_eagerly,
+            # Only `jit_compile` if not eager and in a compatible environment.
+            jit_compile=jit_compile and xla_compatible and not run_eagerly,
+            **kwargs,
+        )
+        # Clear the compiled generate function.
+        self.generate_function = None
+        self.sampler = samplers.get(sampler)
+
+    def make_generate_function(self):
+        """Create or return the compiled generation function."""
+        if self.generate_function is not None:
+            return self.generate_function
+
+        def fn(prompt, input_mask, min_length, max_length):
+            # Create and seed cache with a single forward pass.
+            cache = self.build_cache(prompt)
+
+            def next(prompt, state, index):
+                # The cache index is for our previous token.
+                cache_index = index - 1
+                prompt = tf.slice(prompt, [0, cache_index], [-1, 1])
+                probs, state = self.call_with_cache(prompt, state, cache_index)
+                return tf.squeeze(probs, axis=1), state
+
+            return self.sampler(
+                next=next,
+                prompt=prompt,
+                state=cache,
+                index=min_length,
+                mask=input_mask,
+                end_token_id=self.preprocessor.tokenizer.end_token_id,
+            )
+
+        if self.run_eagerly:
+            self.generate_function = fn
+        else:
+            # `jit_compile` is a property of keras.Model after tf 2.12.
+            # Use `getattr()` for backwards compatibility.
+            jit_compile = getattr(self, "jit_compile", True)
```

Comment:
Would it ever not be set?

Reply:
pre calling `compile()` before tf 2.12, it will not be set IIRC